### ISSUE TITLE
[JOSS Review] Installation issues

### ISSUE THREAD

**Issue by moorepants** [biosustain/optlang#54]
On the first try installing I get a print statement that says "No installers available." on module import:

```
moorepants@garuda:~$ conda create -n optlang-review ipython
Fetching package metadata ...........
Solving package specifications: ..........

Package plan for installation in environment /home/moorepants/miniconda3/envs/optlang-review:

The following NEW packages will be INSTALLED:

    decorator:        4.0.10-py35_1
    ipython:          5.1.0-py35_0 
    ipython_genutils: 0.1.0-py35_0 
    openssl:          1.0.2j-0     
    path.py:          8.2.1-py35_0 
    pexpect:          4.0.1-py35_0 
    pickleshare:      0.7.4-py35_0 
    pip:              9.0.1-py35_1 
    prompt_toolkit:   1.0.9-py35_0 
    ptyprocess:       0.5.1-py35_0 
    pygments:         2.1.3-py35_0 
    python:           3.5.2-0      
    readline:         6.2-2        
    setuptools:       27.2.0-py35_0
    simplegeneric:    0.8.1-py35_1 
    six:              1.10.0-py35_0
    sqlite:           3.13.0-0     
    tk:               8.5.18-0     
    traitlets:        4.3.1-py35_0 
    wcwidth:          0.1.7-py35_0 
    wheel:            0.29.0-py35_0
    xz:               5.2.2-0      
    zlib:             1.2.8-3      

Linking packages ...
[      COMPLETE      ]|############################################################################| 100%
#
# To activate this environment, use:
# > source activate optlang-review
#
# To deactivate this environment, use:
# > source deactivate optlang-review
#

moorepants@garuda:~$ act optlang-review 
(optlang-review) moorepants@garuda:~$ pip install optlang
Collecting optlang
  Using cached optlang-1.0.0-py2.py3-none-any.whl
Requirement already satisfied: six>=1.9.0 in ./miniconda3/envs/optlang-review/lib/python3.5/site-packages (from optlang)
Collecting sympy>=1.0.0 (from optlang)
Collecting mpmath>=0.19 (from sympy>=1.0.0->optlang)
Installing collected packages: mpmath, sympy, optlang
Successfully installed mpmath-0.19 optlang-1.0.0 sympy-1.0
(optlang-review) moorepants@garuda:~$ ipython -c "import optlang"
No solvers available.
(optlang-review) moorepants@garuda:~$ ipython
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) 
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import optlang
No solvers available.

In [2]:                                                                                                  
Do you really want to exit ([y]/n)? 
(optlang-review) moorepants@garuda:~$ python -c "import optlang"
No solvers available.
```

I'd suggest having all required dependencies installed automatically via the pip dependency resolution. Also, shouldn't `import optlang` raise and ImportError if the dependencies are not installed?

**Comment by moorepants** [comment]
After pip installing swiglpk the example runs:

```
(optlang-review) moorepants@garuda:~$ pip install swiglpk
Collecting swiglpk
  Downloading swiglpk-1.3.3-cp35-cp35m-manylinux1_x86_64.whl (1.8MB)
    100% |████████████████████████████████| 1.8MB 549kB/s 
Installing collected packages: swiglpk
Successfully installed swiglpk-1.3.3
(optlang-review) moorepants@garuda:~$ python -c 'import optlang'
(optlang-review) moorepants@garuda:~$ ipython
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) 
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: %paste
from optlang.glpk_interface import Model, Variable, Constraint, Objective

x1 = Variable('x1', lb=0)
x2 = Variable('x2', lb=0)
x3 = Variable('x3', lb=0)

c1 = Constraint(x1 + x2 + x3, ub=100)
c2 = Constraint(10 * x1 + 4 * x2 + 5 * x3, ub=600)
c3 = Constraint(2 * x1 + 2 * x2 + 6 * x3, ub=300)

obj = Objective(10 * x1 + 6 * x2 + 4 * x3, direction='max')

model = Model(name='Simple model')
model.objective = obj
model.add([c1, c2, c3])

status = model.optimize()

## -- End pasted text --

In [2]: print('status:', model.status)
status: optimal

In [3]: print('objective value:', model.objective.value)
objective value: 733.3333333333333

In [4]: for var_name, var in model.variables.iteritems():
   ...:     print(var_name, '=', var.primal)
   ...:     
x1 = 33.33333333333333
x2 = 66.66666666666667
x3 = 0.0
```

The example uses old style Python printing in the README. I'd suggest updating to Python 2.7/3+ valid code.

**Comment by moorepants** [comment]
I see that you have 4 optional dependcies for solvers. I removed swiggplk and tried the scipy backend, but get this error:

```
(optlang-review) moorepants@garuda:~$ pip uninstall swiglpk
Uninstalling swiglpk-1.3.3:
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk-1.3.3.dist-info/DESCRIPTION.rst
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk-1.3.3.dist-info/INSTALLER
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk-1.3.3.dist-info/METADATA
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk-1.3.3.dist-info/RECORD
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk-1.3.3.dist-info/WHEEL
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk-1.3.3.dist-info/metadata.json
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk/.libs/libglpk-190a7631.so.40.1.0
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk/__init__.py
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk/__pycache__/__init__.cpython-35.pyc
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk/__pycache__/swiglpk.cpython-35.pyc
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk/_swiglpk.cpython-35m-x86_64-linux-gnu.so
  /home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/swiglpk/swiglpk.py
Proceed (y/n)? y
  Successfully uninstalled swiglpk-1.3.3
(optlang-review) moorepants@garuda:~$ conda install scipy
Fetching package metadata ...........
Solving package specifications: ..........

Package plan for installation in environment /home/moorepants/miniconda3/envs/optlang-review:

The following NEW packages will be INSTALLED:

    libgfortran: 3.0.0-1           
    mkl:         11.3.3-0          
    numpy:       1.11.2-py35_0     
    scipy:       0.18.1-np111py35_0

Linking packages ...
[      COMPLETE      ]|############################################################################| 100%
(optlang-review) moorepants@garuda:~$ ipython
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) 
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: %paste
from optlang.glpk_interface import Model, Variable, Constraint, Objective

x1 = Variable('x1', lb=0)
x2 = Variable('x2', lb=0)
x3 = Variable('x3', lb=0)

c1 = Constraint(x1 + x2 + x3, ub=100)
c2 = Constraint(10 * x1 + 4 * x2 + 5 * x3, ub=600)
c3 = Constraint(2 * x1 + 2 * x2 + 6 * x3, ub=300)

obj = Objective(10 * x1 + 6 * x2 + 4 * x3, direction='max')

model = Model(name='Simple model')
model.objective = obj
model.add([c1, c2, c3])

status = model.optimize()

## -- End pasted text --
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-32194ddcbb11> in <module>()
----> 1 from optlang.glpk_interface import Model, Variable, Constraint, Objective
      2 
      3 x1 = Variable('x1', lb=0)
      4 x2 = Variable('x2', lb=0)
      5 x3 = Variable('x3', lb=0)

/home/moorepants/miniconda3/envs/optlang-review/lib/python3.5/site-packages/optlang/glpk_interface.py in <module>()
     41 log = logging.getLogger(__name__)
     42 
---> 43 from swiglpk import glp_find_col, glp_get_col_prim, glp_get_col_dual, GLP_CV, GLP_IV, GLP_BV, GLP_UNDEF, GLP_FEAS, \
     44     GLP_INFEAS, GLP_NOFEAS, GLP_OPT, GLP_UNBND, \
     45     glp_set_col_kind, glp_find_row, glp_get_row_prim, glp_get_row_dual, glp_get_obj_val, glp_set_obj_dir, glp_init_smcp, \

ImportError: No module named 'swiglpk'
```

Should any of the backends work?

**Comment by moorepants** [comment]
The installation instructions on readthedocs seem to be more up-to-date and clearer than the README. It would be a good idea to have them match in terms of content.

**Comment by KristianJensen** [comment]
Swiglpk is now a requirement, so after installing optlang with pip, the package can be used for optimizing problems without any further installations.
Since scipy is a huge package, and it's linear programming functionalities are quite limited, we do not think it makes sense to also have this as a mandatory dependency.

The reason that "No solvers available" does not raise an error is that optlang can still be used to formulate, manipulate and save problems even without a solver. With swiglpk as a requirement, this is not so relevant however.

Your comments about the README were very valid and have been addressed (in #55). The example in the README should work with other solvers now.
